android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java

```diff
@@ -26,76 +30,257 @@
+     * if cachedSession is null.
+     *
+     * @param uploadFile    file to upload
+     * @param cachedSession locally cached upload session, or null if not in cache
+     * @return completion of next upload step
+     */
+    @NonNull
+    Completable processUploadForCachedSession(@NonNull UploadFile uploadFile,
+                                              @Nullable UploadSession cachedSession) {
+        checkNotNull(uploadFile, "uploadFile cannot be null");
+
+        Single<UploadSession> sessionSingle;
+        if (cachedSession != null) {
+            // we don't renew expired sessions, because we don't yet know the upload status
+            // successfully completed uploads do not need a new session
+            sessionSingle = Single.just(cachedSession);
+        } else {
+            sessionSingle = getUploadSession(uploadFile).cache();
+        }
+
+        Single<UploadValidationStatus> statusSingle = sessionSingle
+                .flatMap(uploadSession -> getUploadValidationStatus(uploadSession.getId()));
+
+        return Single.zip(
+                Single.just(uploadFile),
+                sessionSingle,
+                statusSingle,
+                this::processUploadForValidationStatus
+        ).flatMapCompletable(i -> i);
     }
 
+    /**
+     * Performs the next upload step, according to the UploadValidationStatus. The normal
+     * transition of UploadValidationStatus is from REQUESTED to VALIDATION_IN_PROGRESS to
+     * REQUESTED.
+     *
+     * @param uploadFile             file being uploaded
+     * @param uploadSession          pre-signed upload session
+     * @param uploadValidationStatus upload validation status from Bridge
+     * @return completion of next upload step
+     */
     @NonNull
-    Single<UploadValidationStatus> uploadToS3(UploadFile uploadFile, UploadSession session) {
+    Completable processUploadForValidationStatus(@NonNull UploadFile uploadFile,
+                                                 @NonNull UploadSession uploadSession,
+                                                 @NonNull UploadValidationStatus uploadValidationStatus) {
+        checkNotNull(uploadFile, "uploadFile cannot be null");
+        checkNotNull(uploadSession, "uploadSession cannot be null");
+        checkNotNull(uploadValidationStatus, "uploadValidationSession cannot be null");
+
+        switch (uploadValidationStatus.getStatus()) {
+            case REQUESTED:
+                return uploadToS3(uploadFile, uploadSession);
+            case SUCCEEDED:
+            case DUPLICATE:
+                return dequeueUpload(uploadFile.filename);
+            case VALIDATION_IN_PROGRESS:
+                LOG.debug("Validation in progress for filename: " + uploadFile.filename +
+                        ", uploadId" + uploadSession.getId());
+                break;
+            case VALIDATION_FAILED:
+                LOG.debug("Validation failed for filename: " + uploadFile.filename +
```

Comment:
Should this dequeueUpload as well?  Or will this fail silently and it will be attempted again?